executor/set.go

```diff
@@ -248,12 +248,33 @@ func (e *SetExecutor) setCharset(cs, co string) error {
 		}
 	}
 	sessionVars := e.ctx.GetSessionVars()
-	for _, v := range variable.SetNamesVariables {
+	if isSetName {
+		for _, v := range variable.SetNamesVariables {
+			if err = sessionVars.SetSystemVar(v, cs); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return errors.Trace(sessionVars.SetSystemVar(variable.CollationConnection, co))
+	}
+	// Set charset statement
+	for _, v := range variable.SetCharsetVariables {
 		if err = sessionVars.SetSystemVar(v, cs); err != nil {
 			return errors.Trace(err)
 		}
 	}
-	return sessionVars.SetSystemVar(variable.CollationConnection, co)
+	csDb, err := sessionVars.GlobalVarsAccessor.GetGlobalSysVar(variable.CharsetDatabase)
+	if err != nil {
+		return err
```

Comment:
Why not using `errors.Trace`?

Reply:
The err return by GetGlobalSysVar already has a stack.